spec/function.dd

```diff
@@ -2952,6 +2960,227 @@ $(H3 $(LNAME2 system-functions, System Functions))
+            $(LI its length is in-bounds with the corresponding memory object,
+            and)
+            $(LI all its elements are safe.)
+        )
+
+        $(P Examples:)
+        $(SPEC_RUNNABLE_EXAMPLE_COMPILE
+        ---
+        int[] f() @system
+        {
+            int[3] a;
+            int[] d1 = a[0 .. 2]; /* d1 is safe. */
+            int[] d2 = a.ptr[0 .. 3]; /* d2 is unsafe because it goes beyond a's
+                bounds. */
+            int*[] d3 = [cast(int*) 0xDEADBEEF]; /* d3 is unsafe because the
+                element is unsafe. */
+            return d1; /* Up to here, d1 was safe, but its pointer becomes
+                invalid when the function returns, so the returned dynamic array
+                is unsafe. */
+        }
+        ---
+        )
+
+        $(P A static array is safe when all its elements are safe. Regardless
+        of the element type, a static array with length zero is always safe.)
+
+        $(P An associative array is safe when all its keys and elements are
+        safe.)
+
+        $(P A struct/union instance is safe when:)
+        $(OL
+            $(LI the values of its accessible fields are safe, and)
+            $(LI it does not introduce $(RELATIVE_LINK2 safe-aliasing, unsafe
+            aliasing) with unions.)
+        )
+
+        $(P Examples:)
+        $(SPEC_RUNNABLE_EXAMPLE_COMPILE
+        ---
+        struct S { int* p; }
+        S s1 = S(new int); /* s1 is safe. */
+        S s2 = S(cast(int*) 0xDEADBEEF); /* s2 is unsafe, because s2.p is
+            unsafe. */
+
+        union U { int* p; size_t x; }
+        U u = U(new int); /* Even though both u.p and u.x are safe, u is unsafe
+            because of unsafe aliasing. */
+        ---
+        )
+
+        $(P A class reference is safe when it is `null` or:)
+        $(OL
+            $(LI it refers to a valid class instance of the stated type or a
+            subtype, and)
+            $(LI the values of the instance's accessible fields are safe, and)
+            $(LI it does not introduce unsafe aliasing with unions.)
+        )
+
+        $(P A function pointer is safe when it is `null` or it refers to a valid
+        function that has the same or a covariant signature.)
+
+        $(P A `delegate` is safe when:)
+        $(OL
+            $(LI its `.funcptr` is `null` or refers to a function that matches
+            the delegate type, and)
+            $(LI its `.ptr` is `null` or refers to a context that is in a format
+            expected by the function.)
+        )
+
+$(H3 $(LNAME2 safe-aliasing, Safe Aliasing))
+
+    $(P When one memory location is accessible with two different types, that
+    aliasing is considered safe in these cases:)
+    $(OL
+        $(LI both types are `const` or `immutable`; or)
+        $(LI one of the types is mutable while the other is a `const`-qualified
+            $(DDSUBLINK spec/type, basic-data-types, basic data type); or)
```

Comment:
It would be nice if there were some general rule that avoided explicitly mentioning basic data types and static arrays with length 0. 
I'm not sure if the list is complete currently, if I interpret it correctly, a `struct Int {int x;}` may not alias with an `int y`, but two raw `int`s can?

Reply:
> It would be nice if there were some general rule that avoided explicitly mentioning basic data types and static arrays with length 0.
> I'm not sure if the list is complete currently,

I think I just gave up at some point with aliasing. I don't think that zero-length arrays are particularly noteworthy, but they're simple to reason about. Structs are hard, with anonymous unions, `align`, and whatnot.

> if I interpret it correctly, a `struct Int {int x;}` may not alias with an `int y`, but two raw `int`s can?

That's right. But the struct case should be added. It even passes `@safe` checks. And at least any aliasing that (correctly) passes as `@safe` should be considered "safe aliasing" by the spec.